Lection 3/task 3.1/ReportApp/Services/Shop/ShopTemplateManagerService.cs

```diff
@@ -0,0 +1,96 @@
+﻿using ClosedXML.Excel;
+using ClosedXML.Report;
+using ReportApp.Models;
+using ReportApp.Models.Shop;
+
+namespace ReportApp.Services.Shop;
+
+public class ShopTemplateManagerService
+{
+    private readonly TemplateManagerService _templateService = new TemplateManagerService();
+
+    private Dictionary<string, Func<ShopReportModel, object>> KeyValuePairs { get; set; } = new Dictionary<string, Func<ShopReportModel, object>>
+    {
+        { "Name", r => r.PointOfPurchase },
+        { "Seller", r => r.Seller },
+        { "Items", r => r.Items }
+    };
+
+    public void FillingAndFormattingExcel(XLTemplate template, ShopReportConfiguration configuration, List<ShopReportModel> models, string type)
+    {
+        var worksheet = template.Workbook.Worksheets.First();
+        worksheet.SetShowGridLines(false);
+
+        var firstDataRow = configuration.DefaultRow;
+        var firstDataColumn = configuration.FirstColumn;
+
+        var groupAmount = 1;
+        var lastDataColumn = configuration.LastColumn;
+        int initialLastRow = configuration.LastRow;
+
+        worksheet.Range(configuration.FirstRowForDynamicGroup, lastDataColumn, configuration.FirstRowForDynamicGroup, configuration.LastColumn)
+            .Delete(XLShiftDeletedCells.ShiftCellsLeft);
+        worksheet.Range(configuration.FirstRowForStaticGroup, lastDataColumn, configuration.FirstRowForStaticGroup, configuration.LastColumn)
+            .Delete(XLShiftDeletedCells.ShiftCellsLeft);
+
+        var titleCell = worksheet.Cell(configuration.ReportTitleRow, firstDataColumn);
+        var style = titleCell.Style;
+        var title = titleCell.Value.ToString();
+
+        var previousRange = worksheet.Range(configuration.ReportTitleRow, firstDataColumn, configuration.ReportTitleRow, configuration.LastColumn).Unmerge();
+        previousRange.Clear();
+
+        var newRange = worksheet.Range(configuration.ReportTitleRow, firstDataColumn, configuration.ReportTitleRow, lastDataColumn - 1).Merge();
+        newRange.Style = style;
+        newRange.Value = title;
+
+        _templateService.CleanTestData(template, configuration, lastDataColumn);
+
+        int currentRow = firstDataRow;
+        int finishedRow = 0;
+
+        foreach (var model in models)
+        {
+            for (int group = 1; group <= groupAmount; group++)
+            {
+                foreach (var item in model.Items)
+                {
+                    int column = firstDataColumn;
+                    foreach (var property in KeyValuePairs)
+                    {
+                        if (property.Key.Equals("Items"))
+                        {
+                            foreach (var prop in item.GetType().GetProperties())
```

Comment:
I love reflections, but why are we using it here? There is much more easy and better way